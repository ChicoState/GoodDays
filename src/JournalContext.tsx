import React, { createContext, useContext, useState } from 'react';
import { JournalEntry } from './journal'; // Assuming the JournalEntry type is defined in './journal'

// Interface for the context type
interface JournalContextType {
  journalList: JournalEntry[];
  loadJournal: (entries: JournalEntry[]) => void;
  addJournalEntry: (entry: JournalEntry) => Promise<void>; // Updated to return a promise
  deleteJournalEntry: (entry: JournalEntry) => void;
}

// Creating the Journal context
export const JournalContext = createContext<JournalContextType | undefined>(undefined);


// Dummy journal entries
const dummyEntries: JournalEntry[] = [
    {
      date: '2024-10-01',
      title: '1st Day',
      hoursActive: 4,
      hoursSleeping: 8,
      hoursFocused: 5,
      hoursOnScreen: 3,
      hoursOutside: 2,
      hoursReading: 1,
      mood: 10,
      reflection: 'Had a productive day!',
      created: new Date().toISOString(),
      updated: new Date().toISOString(),
    },
    {
      date: '2024-10-02',
      title: '2nd Day',
      hoursActive: 3,
      hoursSleeping: 7,
      hoursFocused: 4,
      hoursOnScreen: 4,
      hoursOutside: 1,
      hoursReading: 2,
      mood: 5,
      reflection: 'Just an average day.',
      created: new Date().toISOString(),
      updated: new Date().toISOString(),
    },
    {
      date: '2024-10-03',
      title: '3rd Day',
      hoursActive: 3,
      hoursSleeping: 6.5,
      hoursFocused: 2,
      hoursOnScreen: 5,
      hoursOutside: 0.5,
      hoursReading: 0,
      mood: 4,
      reflection: 'Not that good of a day.',
      created: new Date().toISOString(),
      updated: new Date().toISOString(),
    },
    {
      date: '2024-10-04',
      title: '4th Day',
      hoursActive: 1,
      hoursSleeping: 8,
      hoursFocused: 3,
      hoursOnScreen: 2,
      hoursOutside: 2,
      hoursReading: 0.5,
      mood: 9,
      reflection: 'Liesurely Day.',
      created: new Date().toISOString(),
      updated: new Date().toISOString(),
    },
    {
      date: '2024-10-05',
      title: '5th Day',
      hoursActive: 0.5,
      hoursSleeping: 8,
      hoursFocused: 7,
      hoursOnScreen: 3,
      hoursOutside: 0.5,
      hoursReading: 1,
      mood: 7,
      reflection: 'Focused Day.',
      created: new Date().toISOString(),
      updated: new Date().toISOString(),
    },
  ];


// Hook to use the Journal context

export const useJournal = () => {
  const context = useContext(JournalContext);
  if (!context) {
    throw new Error('useJournal must be used within a JournalProvider');
  }
  return context;
};

// JournalProvider component to wrap the app and provide context
export const JournalProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  // State to hold journal entries
  const [journalEntries, setJournalEntries] = useState<JournalEntry[]>([]); // Initialize with empty array

  // Function to add a new journal entry
  const addJournalEntry = async (entry: JournalEntry) => {
    try {
      // Update the journal entries state
      setJournalEntries(prevEntries => {
        const updatedEntries = [...prevEntries, entry];

        // Save the updated journal entries using Electron's IPC
        window.electron.saveJournalEntry(updatedEntries)
          .then((message: string) => {
            console.log('Journal saved:', message);
          })
          .catch((error: any) => {
            console.error('Error saving journal:', error);
          });

        return updatedEntries; // Return the updated entries array
      });
    } catch (error) {
      console.error('Failed to add journal entry:', error);
    }
  };
  const deleteJournalEntry = (entry: JournalEntry) => {
    setJournalEntries(prevEntries => {
      // Find the index of the entry that matches the title and date
      const index = prevEntries.findIndex(event => event.date === entry.date && event.title === entry.title);

      if (index !== -1) {
        // Create a new array without the matching entry
        const updatedEntries = prevEntries.filter((_, i) => i !== index);

        // Save the updated journal entries using Electron's IPC
        window.electron.saveJournalEntry(updatedEntries)
          .then((message: string) => {
            console.log('Journal updated after deletion:', message);
          })
          .catch((error: any) => {
            console.error('Error saving journal after deletion:', error);
          });

        return updatedEntries; // Update the state with the filtered entries
      } else {
        console.log("Entry not found");
        return prevEntries; // Return the original state if no match
      }
    });
  };

  // Function to load journal entries
  const loadJournal = (entries: JournalEntry[]) => {
    console.log("updating journal entries to new list:", entries);

    if (entries == null) {
      console.warn("attempted to load null entries");
    } else {
      setJournalEntries(entries); // Update state with loaded entries
    }
  };

  // Returning the context provider with the journal list and functions
  return (
    <JournalContext.Provider value={{ journalList: journalEntries, addJournalEntry, loadJournal,  deleteJournalEntry }}>
      {children}
    </JournalContext.Provider>
  );
};
