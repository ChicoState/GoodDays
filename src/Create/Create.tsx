import React, { useState } from "react";
import { useNavigate } from "react-router-dom";

import { JournalEntry } from "../journal";
import { useJournal } from "../JournalContext";

function JournalEntryForm(){

    const { addJournalEntry } = useJournal();

    const [journalEntry, setJournalEntry] = useState<JournalEntry>({
      date: "",
      title: "",
      hoursActive: 0,
      hoursSleeping: 0,
      hoursFocused: 0,
      hoursOnScreen: 0,
      hoursOutside: 0,
      hoursReading: 0,
      mood: 5,
      created: new Date().toISOString(),
      reflection: "",
      updated: null
    });
    
    // for navigation back to home page after submitting
    const navigate = useNavigate();

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
        const { name, value } = event.target;
    
        // update journal on change
        setJournalEntry({
          ...journalEntry,
            [name]: (name.startsWith("hours") || name === "mood") // for hours counts and mood, get value and handle empty string
                ? (value === "" ? "" : Number(value))
                : value,
            updated: new Date().toISOString(), // for update I will need to change this. Currently can only create entries
            created: new Date().toISOString()
        });
    }
    
    // Handle form submission
    const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        addJournalEntry(journalEntry);
        navigate("/");
    };

    return(
      <form onSubmit={handleSubmit}>
            <div>
                <label>Date:</label>
                <input
                    type="date"
                    name="date"
                    value={journalEntry.date}
                    onChange={handleInputChange}
                />
            </div>

            <div>
                <label>Title:</label>
                <input
                    type="text"
                    name="title"
                    value={journalEntry.title}
                    onChange={handleInputChange}
                    placeholder="Journal Title"
                />
            </div>

            <div>
                <label>Hours Active:</label>
                <input
                    type="number"
                    name="hoursActive"
                    value={journalEntry.hoursActive}
                    onChange={handleInputChange}
                />
            </div>
            <div>
                <label>Hours Sleeping:</label>
                <input
                    type="number"
                    name="hoursSleeping"
                    value={journalEntry.hoursSleeping}
                    onChange={handleInputChange}
                />
            </div>
            <div>
                <label>Hours Focused:</label>
                <input
                    type="number"
                    name="hoursFocused"
                    value={journalEntry.hoursFocused}
                    onChange={handleInputChange}
                />
            </div>
            <div>
                <label>Hours on Screen:</label>
                <input
                    type="number"
                    name="hoursOnScreen"
                    value={journalEntry.hoursOnScreen}
                    onChange={handleInputChange}
                />
            </div>
            <div>
                <label>Hours Outside:</label>
                <input
                    type="number"
                    name="hoursOutside"
                    value={journalEntry.hoursOutside}
                    onChange={handleInputChange}
                />
            </div>
            <div>
                <label>Hours Reading:</label>
                <input
                    type="number"
                    name="hoursReading"
                    value={journalEntry.hoursReading}
                    onChange={handleInputChange}
                />
            </div>

            <div>
                <label>Mood (1-10):</label>
                <input
                    type="number"
                    name="mood"
                    min="1"
                    max="10"
                    value={journalEntry.mood}
                    onChange={handleInputChange}
                />
            </div>
        
            <div>
                <label>Contents:</label>
                <textarea
                    name="reflection"
                    value={journalEntry.reflection}
                    onChange={handleInputChange}
                    placeholder="What happened today?"
                />
            </div>

            <button type="submit">Submit Journal Entry</button>

        </form>

    );
}




type CreateProps = {
  //
};


const Create = (props: CreateProps) => {
    return <div>
      <JournalEntryForm/>
    </div>;
}

export default Create;
