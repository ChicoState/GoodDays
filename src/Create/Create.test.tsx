 import * as React from "react";
 import {render, screen} from "@testing-library/react";
import Create from "./Create";
import { JournalProvider } from "../JournalContext";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";

describe("Render Create component", () => {
    test("matches snapshot", () => {
        
        const {asFragment} = render(<JournalProvider><MemoryRouter><Create /></MemoryRouter></JournalProvider>); //fragment is portion of what user sees
        expect(asFragment()).toMatchSnapshot();
    });

    test("Shows proper input fields", () => {
        render (<JournalProvider><MemoryRouter><Create /></MemoryRouter></JournalProvider>);
        //expect(screen.getByRole("textbox", { name: "Date:" })).toBeInTheDocument();
        //expect(screen.getByLabelText(/Date:/i)).toBeInTheDocument();
    });
});