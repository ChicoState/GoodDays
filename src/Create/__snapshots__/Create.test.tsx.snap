// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div>
        <label>
          Date:
        </label>
        <input
          name="date"
          type="date"
          value=""
        />
      </div>
      <div>
        <label>
          Title:
        </label>
        <input
          name="title"
          placeholder="Journal Title"
          type="text"
          value=""
        />
      </div>
      <div>
        <label>
          Hours Active:
        </label>
        <input
          name="hoursActive"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Sleeping:
        </label>
        <input
          name="hoursSleeping"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Focused:
        </label>
        <input
          name="hoursFocused"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours on Screen:
        </label>
        <input
          name="hoursOnScreen"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Outside:
        </label>
        <input
          name="hoursOutside"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Reading:
        </label>
        <input
          name="hoursReading"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Mood (1-10):
        </label>
        <input
          max="10"
          min="1"
          name="mood"
          type="number"
          value="5"
        />
      </div>
      <div>
        <label>
          Contents:
        </label>
        <textarea
          name="reflection"
          placeholder="What happened today?"
        />
      </div>
      <button
        type="submit"
      >
        Submit Journal Entry
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Render Create component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div>
        <label>
          Date:
        </label>
        <input
          name="date"
          type="date"
          value=""
        />
      </div>
      <div>
        <label>
          Title:
        </label>
        <input
          name="title"
          placeholder="Journal Title"
          type="text"
          value=""
        />
      </div>
      <div>
        <label>
          Hours Active:
        </label>
        <input
          name="hoursActive"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Sleeping:
        </label>
        <input
          name="hoursSleeping"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Focused:
        </label>
        <input
          name="hoursFocused"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours on Screen:
        </label>
        <input
          name="hoursOnScreen"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Outside:
        </label>
        <input
          name="hoursOutside"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Hours Reading:
        </label>
        <input
          name="hoursReading"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label>
          Mood (1-10):
        </label>
        <input
          max="10"
          min="1"
          name="mood"
          type="number"
          value="5"
        />
      </div>
      <div>
        <label>
          Contents:
        </label>
        <textarea
          name="reflection"
          placeholder="What happened today?"
        />
      </div>
      <button
        type="submit"
      >
        Submit Journal Entry
      </button>
    </form>
  </div>
</DocumentFragment>
`;
